rsearch/processor.go

```diff
@@ -0,0 +1,112 @@
+// Copyright (c) 2016 Pani Networks
+// All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License. You may obtain
+// a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations
+// under the License.
+
+package rsearch
+
+import (
+	"log"
+)
+
+/*
+{"type":"ADDED","object":{"apiVersion":"romana.io/demo/v1","kind":"NetworkPolicy","metadata":{"name":"pol1","namespace":"default","selfLink":"/apis/romana.io/demo/v1/namespaces/default/networkpolicys/pol1","uid":"d7036130-e119-11e5-aab8-0213e1312dc5","resourceVersion":"119875","creationTimestamp":"2016-03-03T08:28:00Z","labels":{"owner":"t1"}},"spec":{"allowIncoming":{"from":[{"pods":{"tier":"frontend"}}],"toPorts":[{"port":80,"protocol":"TCP"}]},"podSelector":{"tier":"backend"}}}}
+*/
+
+// Process is a goroutine that consumes resource update events and maintains a searchable
+// cache of all known resources. It also accepts search requests and perform searches.
+func Process(in <-chan Event, done chan Done, config Config) chan<- SearchRequest {
+	// Channel to submit SearchRequest's into
+	req := make(chan SearchRequest)
+
+	// storage map is a cache of known KubeObjects
+	// arranged by NPid
+	storage := make(map[string]KubeObject)
+
+	// search map is a cache of known NPid's
+	// arranged by Selectors, where selector being
+	// a field by which we search
+	search := make(map[string]map[string]bool)
+
+	go func() {
+		for {
+			select {
+			case e := <-in:
+				// On incoming event update caches
+				updateStorage(e, storage, search, config)
+			case request := <-req:
+				// On incoming search request return a list
+				// of resources with matching Selectors
+				processSearchRequest(storage, search, request, config)
+			case <-done:
+				return
+			}
+		}
+	}()
+
+	return req
+}
+
+func processSearchRequest(storage map[string]KubeObject, search map[string]map[string]bool, req SearchRequest, config Config) SearchResponse {
+	if config.Server.Debug {
+		log.Println("Received request", req)
+	}
+
+	var resp []KubeObject
+
+	if config.Server.Debug {
+		log.Printf("Index map has following %s, request tag is %s ", search, req.Tag)
+	}
+
+	// Assembling response.
+	for NPid, _ := range search[string(req.Tag)] {
```

Comment:
`Process` goroutine owns all the maps and it can perform only one task at a time.

we either writing or reading, not both at same time
```
                        case e := <-in:
                                // On incoming event update caches
                                updateStorage(e, storage, search, config)
                        case request := <-req:
                                // On incoming search request return a list
                                // of resources with matching Selectors
                                processSearchRequest(storage, search, request, config)
```